wallet-core/src/ffi.rs

```diff
@@ -196,10 +263,215 @@ pub unsafe fn pick_notes(
 
 /// Gets the bookmark from the given note.
 #[no_mangle]
-pub unsafe fn bookmark(leaf_ptr: *const u8, bookmark: *mut u64) -> ErrorCode {
-    let leaf: NoteLeaf = mem::from_buffer(leaf_ptr)?;
+pub unsafe fn bookmarks(
+    notes_ptr: *const u8,
+    bookmarks_ptr: *mut *mut u8,
+) -> ErrorCode {
+    let notes: Vec<NoteLeaf> = mem::from_buffer(notes_ptr)?;
+
+    let bookmarks: Vec<u64> =
+        notes.into_iter().map(|leaf| *leaf.note.pos()).collect();
+
+    let bytes: Vec<u8> = bookmarks
+        .iter()
+        .flat_map(|&num| num.to_le_bytes())
+        .collect();
+
+    let ptr = mem::malloc(bytes.len() as u32);
+    let ptr = ptr as *mut u8;
+
+    *bookmarks_ptr = ptr;
+
+    ptr::copy_nonoverlapping(bytes.as_ptr(), ptr, bytes.len());
+
+    ErrorCode::Ok
+}
+
+#[derive(Default)]
+struct NoOpProver {
+    circuits: core::cell::RefCell<Vec<u8>>,
+}
+
+impl Prove for NoOpProver {
+    fn prove(&self, circuits: &[u8]) -> Result<Vec<u8>, execution_core::Error> {
+        *self.circuits.borrow_mut() = circuits.to_vec();
+
+        Ok(circuits.to_vec())
+    }
+}
+
+#[no_mangle]
+pub unsafe fn into_proved(
+    tx_ptr: *const u8,
+    proof_ptr: *const u8,
+    proved_ptr: *mut *mut u8,
+    hash_ptr: &mut [u8; 64],
+) -> ErrorCode {
+    let tx = mem::read_buffer(tx_ptr);
+    let mut tx: phoenix::Transaction = mem::parse_buffer(tx)?;
```

Comment:
It would be helpful if we can provide an error if the transaction is invalid bytes